dali/operators/reader/loader/webdataset_loader.cc

```diff
@@ -0,0 +1,371 @@
+          spec.GetArgument<std::string>("missing_component_behavior"))) {
+  DALI_ENFORCE(uris_.size() == index_paths_.size(),
+               "Number of uris does not match the number of index files");
+  DALI_ENFORCE(uris_.size() > 0, "No webdataset shards provided");
+  DALI_ENFORCE(missing_component_behavior_ != detail::wds::MissingExtBehavior::Invalid,
+               "Invalid value for missing_component_behavior");
+
+  std::vector<std::string> samples_exts = spec.GetRepeatedArgument<std::string>("ext");
+  ext_.reserve(samples_exts.size());
+
+  // splitting extension bundles by the delimiter
+  for (size_t exts_idx = 0; exts_idx < samples_exts.size(); exts_idx++) {
+    std::stringstream exts_stream(samples_exts[exts_idx]);
+    std::string ext;
+    ext_.emplace_back();
+    while (std::getline(exts_stream, ext, detail::wds::kExtDelim)) {
+      if (!ext_.back().count(ext)) {
+        ext_.back().insert(ext);
+      }
+    }
+  }
+
+  auto dtypes_ids = spec.HasArgument("dtypes") ? spec.GetRepeatedArgument<DALIDataType>("dtypes") :
+                                                 std::vector<DALIDataType>(ext_.size(), DALI_UINT8);
+  dtypes_.reserve(dtypes_ids.size());
+  std::transform(dtypes_ids.begin(), dtypes_ids.end(), std::back_inserter(dtypes_),
+                 TypeTable::GetTypeInfo);
+
+  for (auto& dtype : dtypes_) {
+    DALI_ENFORCE(detail::wds::kSupportedTypes.count(dtype.id()),
+                 make_string("Unsupported output dtype ", dtype.name(),
+                             ". Supported types include: ", SupportedTypesListGen()));
+  }
+  DALI_ENFORCE(ext_.size() == dtypes_.size(),
+               "Number of extensions does not match the number of types");
+}
+
+WebdatasetLoader::~WebdatasetLoader() {}
+
+void WebdatasetLoader::PrepareEmpty(vector<Tensor<CPUBackend>>& empty) {
+  empty = std::vector<Tensor<CPUBackend>>(ext_.size());
+  for (size_t output_index = 0; output_index < ext_.size(); output_index++) {
+    empty[output_index].reserve(tensor_init_bytes_);
+    empty[output_index].set_type(dtypes_[output_index]);
+  }
+}
+
+inline std::string GetExtension(const std::string& filepath) {
+  const size_t dot_pos = filepath.find_first_of('.', filepath.find_last_of('/') + 1);
+  return filepath.substr(dot_pos + 1);
+}
+
+void WebdatasetLoader::ReadSample(vector<Tensor<CPUBackend>>& sample) {
+  MoveToNextShard(sample_index_);
+  detail::wds::SampleDesc& current_sample = samples_[sample_index_];
+  auto& current_wds_shard = wds_shards_[current_sample.wds_shard_index];
+
+  for (auto& component : current_sample.components) {
+    current_wds_shard.SeekArchive(component.offset);
+
+    // Checking if the component data from the index file agrees with reality
+    const auto& index_path = index_paths_[current_sample.wds_shard_index];
+    DALI_ENFORCE(
+        !current_wds_shard.EndOfArchive(),
+        IndexFileErrMsg(index_path, current_sample.line_number, "archive shorter than reported"));
```

Comment:
```suggestion
        IndexFileErrMsg(index_path, current_sample.line_number, "archive shorter than indicated in corresponding index"));
```

Reply:
But the index and line number is being printed already, does the user need a further indication that the error is regarding the index file?

Reply:
Then use something like:
```suggestion
        IndexFileErrMsg(index_path, current_sample.line_number, "offset is outside of the archive file."));
```